### PR TITLE
WIP: Fixes problems with TimeStepManager

### LINKED ISSUE #821: TimeStepManager and transport
TimeStepManager stores a "dt_stable_storage", which is intended for use in the following situation:

1. Flow admits large dt, let's say 10 days.
2. Events (e.g. checkpoint request) limits this to let's say 1 day.
3. In the next get_dt() call, because flow only successfully took a 1 day step, it now "votes" for 1 day.
4. Then, in the TimeStepManaager, if the last step was limited by an event, rather than listening to what flow actually voted for (1 day) we use dt_stable_storage = 10 days.

This is needed to recover the "good" step size because Flow PK isn't smart enough to realize that despite only taking a 1 day tilmestep, it could happily have taken a 10 day step.

This breaks in the case of transport.  Currently, I am seeing the following:

1. Transport's initial step has 0 flux, so CFL has no limitation, so the initial dt provided by transport is "initial time step [s]" provided by the user.  Let's say this is large, e.g. 10 days.
2. Events limit this to 1 day.
3. In the next get_dt() call, transport's CFL kicks in and votes for 1 hour.
4. TimeStepManager thinks it knows that transport can take a 10 day step, and overrides the 1 hour vote by transport in favor of 10 days.

In the test I'm debugging, this gets chained -- every single step is limited by the 1 day checkpoint event, so every single step transport votes for 1 hour due to CFL, which is then overwritten by dt_stable_storage to 10 days, which is then limited to 1 day, and every single step is 1 day (bigger than transport's CFL).

This is more problematic version of the (unfixed) problem reported here: https://github.com/amanzi/amanzi/issues/685#issuecomment-1289085020 

I would like to change the code to do the following:

1. remove dt_stable_storage from the TimeStepManager
2. Let each PK keep track of both the dt it requested and the dt it received from the driver -- if it was limited and it thinks it can keep the bigger dt, then it should do that itself.

This would fix both problems.

The end result is that

`TimeStepManager::TimeStep(T, dT, after_failure) {}`

should _never_ return a value bigger than dT.

### PR DESCRIPTION
**Pull request by ecoon**
closes #821 and #685

Note this definitely broke some ATS regression tests.

It may break some Amanzi regression tests as well, due to differing timestep history.

### CODE REVIEW

**Comment by ecoon** [comment]
@lipnikov  Note that this may require some changes to Amanzi PKs.  See, for example, my changes to Amanzi's Richards_PK.cc for an example of what I mean.

Specifically, if a PK recommends a large step size (say 1 day), and an event limits that size to a small step size (say, 1 hour).  Then, the PK may succeed, and even grow the step size (to maybe 1.5 hours), but then future steps after the event are also limited to the reduce step size (1.5 hours) and may not grow back to the valid PK step size (1 day).  This is due to the elimination of dt_stable_size from TimeStepManager (for reasons in the bugs referenced above).  I have provided an example fix for Richards PK which should work, conceptually, in other PKs as well.  Effectively the PKs need to keep a record of the step size they recommended and, if the actual step size taken is smaller than that recommendation and all goes well with the solve, then one should not reduce the recommended step size for the next step.

**Comment by lipnikov** [comment]
@ecoon Time step controllers were designed to help a PK in tracking time steps. I am trying to understand why we still need to do time step management inside a PK. The logic regarding time events is quite generic to be moved to a controller. Thoughts?

File: src/utils/Event.hh [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  {
+    if (period_ <= (Scalar)0) {
+      Errors::Message msg("EventSPS provided invalid period.");
+      Exceptions::amanzi_throw(msg);
+    }
+    if (stop_ > (Scalar)0 && stop_ <= start_) {
+      Errors::Message msg("EventSPS provided invalid start/stop.");
+      Exceptions::amanzi_throw(msg);
+    }
+    if (start_ < (Scalar)0) {
```
Comment (lipnikov) [inline src/utils/Event.hh:106]: It would be good for a supporting capability to not assume anything about time flow in events or physics. I understand that it is probably overkill for geophysics applications.

File: src/utils/Event.hh [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+// Either int or double
+//
+template<typename Scalar>
+struct Event_EPS {
+  static constexpr int value = 0;
+};
+
+template<>
+struct Event_EPS<double> {
+  static constexpr double value = 1.e-12;
```
Comment (lipnikov) [inline src/utils/Event.hh:30]: Is this a physics-based constraint or a numerical round-off constraint? If the first one, we probably need a way to set it up. If the second one, why is it so large?

File: src/utils/TimeStepManager.cc [hunk 3]
```diff
@@ -16,109 +16,108 @@
...
 }
 
-TimeStepManager::TimeStepManager(Teuchos::ParameterList& verb_list)
+TimeStepManager::TimeStepManager(Teuchos::ParameterList& plist)
+  : manual_override_(false)
 {
-  dt_stable_storage = -1.;
-  vo_ = Teuchos::rcp(new VerboseObject("TimeStepManager", verb_list));
+  // manual override
+  if (plist.isParameter("prescribed timesteps [s]")) {
```
Comment (lipnikov) [inline src/utils/TimeStepManager.cc:39]: SI is the default unit system in Amanzi. Hence, we should use unit only for exceptions.

File: src/utils/test/utils_time_step_manager.cc [hunk 4]
```diff
@@ -16,7 +16,7 @@
 
 TEST(TIME_STEP_MANAGER)
```
Comment (lipnikov) [inline src/utils/test/utils_time_step_manager.cc:17]: Many problems we had before were related to small dt. Do we have tests for such events?

**Comment by ecoon** [comment]
> @ecoon Time step controllers were designed to help a PK in tracking time steps. I am trying to understand why we still need to do time step management inside a PK. 

That's probably right, we could pretty straightforwardly move the "stable step" size into the TimestepController, and as long as all PKs use the TimeStepController class, it should be fine.  I'll think about this a bit.

> The logic regarding time events is quite generic to be moved to a controller. Thoughts?

Events are outside of PKs -- they only need to live in one place (the cycle driver).

But I agree that if an event limits the step size, probably the logic to ignore that reduction in step size can be moved into the TimeStepController rather than in the PK.

File: src/utils/test/utils_time_step_manager.cc (hunk 4 above)
Comment (ecoon) [reply src/utils/test/utils_time_step_manager.cc:17]: I think we have a test that checks if a dt is 0.99 * the time remaining, that we instead take 1/2 the time remaining.

I'm not sure we have much else.  I'm happy to start adding more tests in utils_time_step_manager.cc if you have ideas of corner cases that have broken.  I could mock up a test that demonstrates a few common failures and add them as well.

File: src/utils/TimeStepManager.cc (hunk 3 above)
Comment (ecoon) [reply src/utils/TimeStepManager.cc:39]: I'll add both -- I've taken to preferring to use the unit everywhere because it makes the input spec self-documenting.  It's easy enough to tell people "all things are SI unless otherwise specified" but its even easier to not tell them and have it in the name.

Eventually I think it would be good to stop using ParameterList entirely and expand upon Teuchos ParameterList to include units (e.g. `<Parameter name="..." unit="..." value="..."/>` ) in the native spec.

File: src/utils/Event.hh (hunk 2 above)
Comment (ecoon) [reply src/utils/Event.hh:30]: It's what was in the code previously.  I think it's a numerical round-off contraint.  It could probably be a few orders smaller, I haven't tried.

File: src/utils/Event.hh (hunk 1 above)
Comment (ecoon) [reply src/utils/Event.hh:106]: You're saying negative start times are ok with you?  I'm ok with that.

**Comment by ecoon** [comment]
Note this also closes #865
